ml-agents-envs/mlagents_envs/rpc_communicator.py

```diff
@@ -86,22 +86,39 @@ def check_port(self, port):
         finally:
             s.close()
 
-    def poll_for_timeout(self):
+    def poll_for_timeout(self, poll_callback: Optional[PollCallback] = None) -> None:
         """
         Polls the GRPC parent connection for data, to be used before calling recv.  This prevents
         us from hanging indefinitely in the case where the environment process has died or was not
         launched.
-        """
-        if not self.unity_to_external.parent_conn.poll(self.timeout_wait):
-            raise UnityTimeOutException(
-                "The Unity environment took too long to respond. Make sure that :\n"
-                "\t The environment does not need user interaction to launch\n"
-                '\t The Agents\' Behavior Parameters > Behavior Type is set to "Default"\n'
-                "\t The environment and the Python interface have compatible versions."
-            )
 
-    def initialize(self, inputs: UnityInputProto) -> UnityOutputProto:
-        self.poll_for_timeout()
+        Additionally, a callback can be passed to periodically check the state of the environment.
+        This is used to detect the case when the environment dies without cleaning up the connection,
+        so that we can stop sooner and raise a more appropriate error.
+        """
+        wait_time_remaining = self.timeout_wait
+        callback_timeout_wait = self.timeout_wait // 10
```

Comment:
Could make the "10" configurable, but seemed like a good rule of thumb.

Reply:
Don't really care if the timeout is configurable, but if you do make it configurable please make sure to check it is less than `self.timeout_wait`